File: lazypx4/jobs.py
"""Generic background subprocess runner for one-shot utility actions.

Firmware flashing (``px_uploader.py``), the .ulog web upload
(``upload_log.py``) and the ecl_ekf health-check
(``process_logdata_ekf.py``) all just run a standalone script to completion
and report a result - none of them are something the vehicle link depends
on, and only one is ever meaningfully in flight at a time. Routing all three
through one job runner (and one ``state.job_*`` slot, see :mod:`lazypx4.state`)
keeps their screens' code identical instead of three near-duplicate copies of
the same "run a subprocess, stream its output, show a status line" logic.
"""

from __future__ import annotations

import subprocess
import threading
import time

from .eventlog import log_command, log_error, log_info, log_warn
from .state import state

_ACTIVE_PROC = None
_ACTIVE_PROC_LOCK = threading.Lock()


def job_running():
    with state.lock:
        return state.job_active


def start_job(kind, cmd, label, on_success=None, cwd=None):
    """Run ``cmd`` (an argv list) in the background under ``label``.

    ``on_success(output_text)`` runs on the worker thread once the process
    exits with status 0, and its return value (if any) becomes
    ``state.job_result``. Raising from it fails the job with that exception's
    message - used by callers that only know the run truly succeeded once
    they can parse the process's own output (e.g. the uploaded log's URL).
    """
    with state.lock:
        if state.job_active:
            log_warn(f"{label}: another background job is already running")
            return False

        state.job_active = True
        state.job_kind = kind
        state.job_cancel = False
        state.job_lines.clear()
        state.job_status = "RUNNING"
        state.job_error = ""
        state.job_result = ""
        state.job_started_at = time.monotonic()

    threading.Thread(
        target=_run_job, args=(cmd, on_success, cwd, label),
        daemon=True, name=f"Job-{kind}",
    ).start()

    log_command(f"{label} started")
    return True


def cancel_job():
    with state.lock:
        if not state.job_active:
            return False
        state.job_cancel = True

    with _ACTIVE_PROC_LOCK:
        proc = _ACTIVE_PROC
    if proc is not None:
        try:
            proc.terminate()
        except Exception:
            pass

    log_command("Job cancellation requested")
    return True


def _append_line(line):
    if not line:
        return
    with state.lock:
        state.job_lines.append(line)


def _run_job(cmd, on_success, cwd, label):
    global _ACTIVE_PROC

    output_lines = []

    try:
        proc = subprocess.Popen(
            cmd, cwd=cwd,
            stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
        )
    except Exception as exc:
        with state.lock:
            state.job_active = False
            state.job_status = "ERROR"
            state.job_error = str(exc)
        log_error(f"{label} failed to start: {exc}")
        return

    with _ACTIVE_PROC_LOCK:
        _ACTIVE_PROC = proc

    try:
        buf = bytearray()
        while True:
            chunk = proc.stdout.read(1)
            if not chunk:
                break
            if chunk in (b"\r", b"\n"):
                line = buf.decode("utf-8", errors="replace").strip()
                buf.clear()
                if line:
                    _append_line(line)
                    output_lines.append(line)
            else:
                buf.extend(chunk)

        if buf:
            line = buf.decode("utf-8", errors="replace").strip()
            if line:
                _append_line(line)
                output_lines.append(line)

        returncode = proc.wait(timeout=15)
    except Exception as exc:
        try:
            proc.kill()
        except Exception:
            pass
        with state.lock:
            state.job_active = False
            state.job_status = "ERROR"
            state.job_error = str(exc)
        log_error(f"{label} failed: {exc}")
        return
    finally:
        with _ACTIVE_PROC_LOCK:
            _ACTIVE_PROC = None

    with state.lock:
        cancelled = state.job_cancel

    if cancelled:
        with state.lock:
            state.job_active = False
            state.job_status = "CANCELLED"
        log_warn(f"{label} cancelled")
        return

    if returncode != 0:
        error_text = output_lines[-1] if output_lines else f"exit code {returncode}"
        with state.lock:
            state.job_active = False
            state.job_status = "ERROR"
            state.job_error = error_text
        log_error(f"{label} failed: {error_text}")
        return

    full_output = "\n".join(output_lines)
    result_text = ""

    if on_success is not None:
        try:
            result_text = on_success(full_output) or ""
        except Exception as exc:
            with state.lock:
                state.job_active = False
                state.job_status = "ERROR"
                state.job_error = str(exc)
            log_error(f"{label}: {exc}")
            return

    with state.lock:
        state.job_active = False
        state.job_status = "SUCCESS"
        state.job_result = result_text

    log_info(f"{label}: done" + (f" - {result_text}" if result_text else ""))

File: lazypx4-rs/src/jobs.rs
//! Background subprocess jobs: firmware flashing (`px_uploader.py`), the
//! .ulog web upload (`upload_log.py`) and the ecl_ekf health check
//! (`process_logdata_ekf.py`). Mirrors `jobs.py` + `pxtools.py`.
//!
//! These stay real PX4 tool scripts run under Python - they're manual,
//! occasional actions, and reimplementing a bootloader protocol or the
//! ecl_ekf analysis buys nothing. One job slot is shared by all three.

use std::collections::VecDeque;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::process::{Child, Command, Stdio};
use std::sync::{Arc, Mutex};

use crate::host::lock;
use crate::state::{self, Shared};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum JobKind {
    Firmware,
    UlogUpload,
    EclEkf,
}

#[derive(Debug, Default)]
pub struct JobState {
    pub kind: Option<JobKind>,
    pub active: bool,
    pub cancel: bool,
    pub status: String,
    pub error: String,
    pub result: String,
    pub lines: VecDeque<String>,
    /// Running process, for cancel (the worker owns the Child itself).
    pid: Option<u32>,
}

pub type Jobs = Arc<Mutex<JobState>>;

/// Called with the job's full output once it exited 0; its Ok value becomes
/// the result line, an Err fails the job.
type OnSuccess = Box<dyn FnOnce(&str, &Shared) -> Result<String, String> + Send>;

pub fn running(jobs: &Jobs) -> bool {
    lock(jobs).active
}

/// The interpreter for a Tools/ script: an active virtualenv, the repo's
/// `.venv` next to the tools directory (where lazypx4's `tools` extra puts
/// pyserial / requests / pyulog), else `python3` from PATH.
fn tool_python(tools_dir: &str) -> PathBuf {
    if let Ok(venv) = std::env::var("VIRTUAL_ENV") {
        let p = Path::new(&venv).join("bin/python");
        if p.exists() {
            return p;
        }
    }
    let repo_venv = Path::new(tools_dir).join("../.venv/bin/python");
    if repo_venv.exists() {
        return repo_venv;
    }
    PathBuf::from("python3")
}

fn start(jobs: &Jobs, shared: &Shared, kind: JobKind, mut cmd: Command, label: String, on_success: Option<OnSuccess>) -> bool {
    {
        let mut j = lock(jobs);
        if j.active {
            state::lock(shared).warn(format!("{label}: another background job is already running"));
            return false;
        }
        *j = JobState { kind: Some(kind), active: true, status: "RUNNING".into(), ..Default::default() };
    }

    let child = match cmd.stdin(Stdio::null()).stdout(Stdio::piped()).stderr(Stdio::piped()).spawn() {
        Ok(c) => c,
        Err(e) => {
            let mut j = lock(jobs);
            j.active = false;
            j.status = "ERROR".into();
            j.error = e.to_string();
            state::lock(shared).error(format!("{label} failed to start: {e}"));
            return false;
        }
    };
    lock(jobs).pid = Some(child.id());
    state::lock(shared).command(format!("{label} started"));

    let (jobs, shared) = (jobs.clone(), shared.clone());
    std::thread::spawn(move || run(jobs, shared, child, label, on_success));
    true
}

fn run(jobs: Jobs, shared: Shared, mut child: Child, label: String, on_success: Option<OnSuccess>) {
    let (stdout, stderr) = (child.stdout.take(), child.stderr.take());
    // stderr is merged line-by-line alongside stdout, like STDOUT=STDERR.
    let err_jobs = jobs.clone();
    let err_thread = stderr.map(|s| std::thread::spawn(move || pump(s, &err_jobs)));
    let mut output = stdout.map(|s| pump(s, &jobs)).unwrap_or_default();
    if let Some(t) = err_thread {
        output.extend(t.join().unwrap_or_default());
    }
    let status = child.wait();

    let finish = |status: &str, error: String, result: String| {
        let mut j = lock(&jobs);
        j.active = false;
        j.pid = None;
        j.status = status.into();
        j.error = error;
        j.result = result;
    };

    if lock(&jobs).cancel {
        finish("CANCELLED", String::new(), String::new());
        state::lock(&shared).warn(format!("{label} cancelled"));
        return;
    }
    match status {
        Ok(s) if s.success() => {}
        other => {
            let error = output.last().cloned().unwrap_or_else(|| match other {
                Ok(s) => format!("exit code {}", s.code().unwrap_or(-1)),
                Err(e) => e.to_string(),
            });
            state::lock(&shared).error(format!("{label} failed: {error}"));
            finish("ERROR", error, String::new());
            return;
        }
    }
    let full = output.join("\n");
    let result = match on_success {
        None => Ok(String::new()),
        Some(f) => f(&full, &shared),
    };
    match result {
        Ok(result) => {
            let suffix = if result.is_empty() { String::new() } else { format!(" - {result}") };
            state::lock(&shared).info(format!("{label}: done{suffix}"));
            finish("SUCCESS", String::new(), result);
        }
        Err(e) => {
            state::lock(&shared).error(format!("{label}: {e}"));
            finish("ERROR", e, String::new());
        }
    }
}

/// Read a pipe to EOF, splitting on CR as well as LF (px_uploader draws its
/// progress bar with CR), appending each line to the job's live tail.
fn pump(mut reader: impl Read, jobs: &Jobs) -> Vec<String> {
    let mut all = Vec::new();
    let mut buf = Vec::new();
    let mut chunk = [0u8; 1024];
    let emit = |buf: &mut Vec<u8>, all: &mut Vec<String>| {
        let line = String::from_utf8_lossy(buf).trim().to_string();
        buf.clear();
        if !line.is_empty() {
            let mut j = lock(jobs);
            if j.lines.len() >= 300 {
                j.lines.pop_front();
            }
            j.lines.push_back(line.clone());
            all.push(line);
        }
    };
    while let Ok(n) = reader.read(&mut chunk) {
        if n == 0 {
            break;
        }
        for &b in &chunk[..n] {
            if b == b'\n' || b == b'\r' {
                emit(&mut buf, &mut all);
            } else {
                buf.push(b);
            }
        }
    }
    emit(&mut buf, &mut all);
    all
}

pub fn cancel(jobs: &Jobs, shared: &Shared) -> bool {
    let pid = {
        let mut j = lock(jobs);
        if !j.active {
            return false;
        }
        j.cancel = true;
        j.pid
    };
    if let Some(pid) = pid {
        // SAFETY: plain signal to the child we spawned; it is only reaped
        // by the worker's wait(), after which pid is cleared.
        unsafe {
            libc::kill(pid as libc::pid_t, libc::SIGTERM);
        }
    }
    state::lock(shared).command("Job cancellation requested");
    true
}

// ---------------------------------------------------------------------------
// Firmware flashing
// ---------------------------------------------------------------------------

/// `.px4` files in the firmware directory, newest first.
pub fn discover_firmware_files(dir: &str) -> Vec<PathBuf> {
    let Ok(entries) = std::fs::read_dir(dir) else { return Vec::new() };
    let mut files: Vec<(std::time::SystemTime, PathBuf)> = entries
        .filter_map(Result::ok)
        .map(|e| e.path())
        .filter(|p| p.is_file() && p.extension().is_some_and(|e| e == "px4"))
        .map(|p| (p.metadata().and_then(|m| m.modified()).unwrap_or(std::time::UNIX_EPOCH), p))
        .collect();
    files.sort_by_key(|f| std::cmp::Reverse(f.0));
    files.into_iter().map(|(_, p)| p).collect()
}

/// USB serial ports a flight controller could be on (never a PC's onboard
/// /dev/ttyS* UARTs).
pub fn discover_serial_ports() -> Vec<String> {
    let mut ports: Vec<String> = Vec::new();
    for (dir, prefixes) in [("/dev/serial/by-id", &[""][..]), ("/dev", &["ttyACM", "ttyUSB"][..])] {
        if let Ok(entries) = std::fs::read_dir(dir) {
            for e in entries.filter_map(Result::ok) {
                let name = e.file_name().to_string_lossy().into_owned();
                if prefixes.iter().any(|p| name.starts_with(p)) {
                    ports.push(format!("{dir}/{name}"));
                }
            }
        }
    }
    ports.sort();
    ports.dedup();
    ports
}

pub fn start_firmware_flash(jobs: &Jobs, shared: &Shared, tools_dir: &str, firmware: &Path, port: &str) -> bool {
    let uploader = Path::new(tools_dir).join("px_uploader.py");
    if !uploader.is_file() {
        state::lock(shared).error(format!("px_uploader.py not found under {tools_dir}"));
        return false;
    }
    let mut cmd = Command::new(tool_python(tools_dir));
    cmd.arg(&uploader).arg("--port").arg(port).arg(firmware);
    let name = firmware.file_name().unwrap_or_default().to_string_lossy();
    start(jobs, shared, JobKind::Firmware, cmd, format!("Firmware flash ({name} -> {port})"), None)
}

// ---------------------------------------------------------------------------
// .ulog upload / ecl_ekf check
// ---------------------------------------------------------------------------

fn git_email() -> String {
    Command::new("git")
        .args(["config", "--global", "user.email"])
        .stderr(Stdio::null())
        .output()
        .ok()
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .unwrap_or_default()
}

/// Best-effort copy to the clipboard: wl-copy (Wayland), else xclip.
fn copy_to_clipboard(text: &str) -> bool {
    for (bin, args) in [("wl-copy", &[][..]), ("xclip", &["-selection", "clipboard"][..])] {
        let Ok(mut child) = Command::new(bin).args(args).stdin(Stdio::piped()).stdout(Stdio::null()).stderr(Stdio::null()).spawn() else {
            continue;
        };
        if let Some(mut stdin) = child.stdin.take() {
            use std::io::Write;
            let _ = stdin.write_all(text.as_bytes());
        }
        if child.wait().is_ok_and(|s| s.success()) {
            return true;
        }
    }
    false
}

pub fn start_ulog_upload(jobs: &Jobs, shared: &Shared, tools_dir: &str, server: &str, log: &Path) -> bool {
    let script = Path::new(tools_dir).join("upload_log.py");
    if !script.is_file() {
        state::lock(shared).error(format!("upload_log.py not found under {tools_dir}"));
        return false;
    }
    let mut cmd = Command::new(tool_python(tools_dir));
    cmd.arg(&script)
        .args(["--quiet", "--server", server, "--description", "", "--feedback", "", "--source", "lazypx4"]);
    let email = git_email();
    if !email.is_empty() {
        cmd.args(["--email", &email]);
    }
    cmd.arg(log);

    let on_success: OnSuccess = Box::new(|output, shared| {
        let url = output
            .split("URL:")
            .nth(1)
            .and_then(|rest| rest.split_whitespace().next())
            .ok_or("upload_log.py did not report a plot URL")?
            .to_string();
        if copy_to_clipboard(&url) {
            state::lock(shared).info(format!("Log report URL copied to clipboard: {url}"));
        } else {
            state::lock(shared).warn("Could not copy the URL to the clipboard (wl-copy / xclip)");
        }
        Ok(url)
    });
    let name = log.file_name().unwrap_or_default().to_string_lossy();
    start(jobs, shared, JobKind::UlogUpload, cmd, format!("Uploading {name} to {server}"), Some(on_success))
}

pub fn start_ecl_ekf_check(jobs: &Jobs, shared: &Shared, tools_dir: &str, log: &Path) -> bool {
    let script = Path::new(tools_dir).join("ecl_ekf/process_logdata_ekf.py");
    if !script.is_file() {
        state::lock(shared).error(format!("process_logdata_ekf.py not found under {tools_dir}"));
        return false;
    }
    let mut cmd = Command::new(tool_python(tools_dir));
    cmd.arg(&script).arg(log);
    let report = format!("{}-0.pdf", log.display());
    let on_success: OnSuccess = Box::new(move |output, _| {
        let mut verdict = if output.contains("Minor anomalies detected") {
            "WARNING - minor anomalies detected".to_string()
        } else if output.contains("No anomalies detected") {
            "PASS - no anomalies detected".to_string()
        } else {
            "completed".to_string()
        };
        if Path::new(&report).is_file() {
            verdict += &format!(" - report: {report}");
        }
        Ok(verdict)
    });
    let name = log.file_name().unwrap_or_default().to_string_lossy();
    start(jobs, shared, JobKind::EclEkf, cmd, format!("EKF health-check on {name}"), Some(on_success))
}
